system-addon/content-src/components/TopSites/TopSite.jsx

```diff
@@ -175,24 +239,159 @@ export class TopSite extends React.PureComponent {
     </TopSiteLink>);
   }
 }
-TopSite.defaultProps = {link: {}};
+TopSite.defaultProps = {
+  link: {},
+  onDragStart() {}
+};
 
-export const TopSitePlaceholder = () => <TopSiteLink className="placeholder" />;
+export const TopSitePlaceholder = props => <TopSiteLink className="placeholder" isDraggable={false} {...props} />;
 
-export const TopSiteList = props => {
-  const topSites = props.TopSites.rows.slice(0, props.TopSitesCount);
-  const topSitesUI = [];
-  for (let i = 0, l = props.TopSitesCount; i < l; i++) {
-    const link = topSites[i];
-    topSitesUI.push(!link ? <TopSitePlaceholder key={i} /> : <TopSite
-      key={link.guid || link.url}
-      dispatch={props.dispatch}
-      link={link}
-      index={i}
-      intl={props.intl}
-      onEdit={props.onEdit} />);
-  }
-  return (<ul className="top-sites-list">
-    {topSitesUI}
-  </ul>);
-};
+export class TopSiteList extends React.PureComponent {
+  constructor(props) {
+    super(props);
+    this.state = this.DEFAULT_STATE = {
+      draggedIndex: null,
+      draggedSite: null,
+      draggedTitle: null,
+      topSitesPreview: null
+    };
+    this.onDragEvent = this.onDragEvent.bind(this);
+  }
+  componentWillUpdate(nextProps) {
+    if (this.state.draggedSite) {
+      const prevTopSites = this.props.TopSites && this.props.TopSites.rows;
+      const newTopSites = nextProps.TopSites && nextProps.TopSites.rows;
+      if (prevTopSites && prevTopSites[this.state.draggedIndex] &&
+        prevTopSites[this.state.draggedIndex].url === this.state.draggedSite.url &&
+        (!newTopSites[this.state.draggedIndex] || newTopSites[this.state.draggedIndex].url !== this.state.draggedSite.url)) {
+        // We got the new order from the redux store via props. We can clear state now.
+        this.setState(this.DEFAULT_STATE);
+      }
+    }
+  }
+  userEvent(event, index) {
+    this.props.dispatch(ac.UserEvent({
+      event,
+      source: TOP_SITES_SOURCE,
+      action_position: index
+    }));
+  }
+  onDragEvent(event, index, link, title) {
+    switch (event.type) {
+      case "dragstart":
+        this.onDragStart(index, link, title);
+        break;
+      case "dragend":
+        this.onDragEnd();
+        break;
+      case "dragover":
+      case "dragenter":
+        this.onDragEnter(index);
+        break;
+      case "dragleave":
+        this.onDragLeave();
+        break;
+      case "drop":
+        this.onDrop(index);
```

Comment:
nit: these all seem to be inline-able? Unless we want to expose them for slightly easier testing instead of passing a dummy event into this method?